common/flatpak-run.c

```diff
@@ -3092,8 +3092,94 @@ flatpak_run_add_app_info_args (GPtrArray      *argv_array,
   return TRUE;
 }
 
+static gboolean
+symlink_target_is_deployed (GFile *runtime_files, const char *symlink, const char *target)
+{
+  const char *runtime_files_path = NULL;
+  g_autofree char *deployed_target = NULL;
+
+  /* Manually check whether the symlink would point to a resolvable path inside the sandbox. */
+  runtime_files_path = flatpak_file_get_path_cached (runtime_files);
+  if (target[0] == '/')
+    {
+      /* Symlink points to an absolute path */
+      guint base_index = (g_strcmp0 (target, "/usr") == 0) ? 4 : 1;
+      deployed_target = g_strdup_printf ("%s/%s", runtime_files_path, &target[base_index]);
+    }
+  else
+    {
+      g_autofree char *symlink_subdir = NULL;
+      symlink_subdir = g_path_get_dirname (symlink);
+
+      if (!g_str_has_prefix (target, "../"))
+        {
+          /* Symlink points to a file relative to its base directory */
+          deployed_target = g_strdup_printf ("%s/%s/%s", runtime_files_path, symlink_subdir, target);
+        }
+      else
+        {
+          /* Symlink points to a file relative to and ancestor of its base directory */
+          g_autofree char *target_subdir = NULL;
+          g_autofree char *target_root_subdir = NULL;
+          g_autofree char *target_basename = NULL;
+          g_auto(GStrv) symlink_subdirs = NULL;
+          g_auto(GStrv) target_subdirs = NULL;
+          const char *target_subdir_last_part = NULL;
+          guint n_symlink_path_elems;
+          guint n_target_path_elems;
+
+          target_subdir = g_path_get_dirname (target);
+
+          symlink_subdirs = g_strsplit (symlink_subdir, "/", -1);
+          n_symlink_path_elems = g_strv_length (symlink_subdirs);
+
+          target_subdirs = g_strsplit (target_subdir, "../", -1);
+          n_target_path_elems = g_strv_length (target_subdirs);
```

Comment:
Drop all this fragile path operations too, and rely on GFile and its operations and canonicalization like the above.